components/test_raftstore/src/pd.rs

```diff
@@ -799,6 +904,77 @@ impl TestPdClient {
         self.schedule_operator(region_id, op);
     }
 
+    pub fn joint_confchange(
+        &self,
+        region_id: u64,
+        mut changes: Vec<(ConfChangeType, metapb::Peer)>,
+    ) -> (Vec<metapb::Peer>, Vec<metapb::Peer>) {
+        let region = block_on(self.get_region_by_id(region_id)).unwrap().unwrap();
+        let (mut add_peers, mut remove_peers) = (Vec::new(), Vec::new());
+
+        let to_add_peers = changes
+            .iter()
+            .filter(|(c, _)| *c != ConfChangeType::RemoveNode)
+            .map(|(_, p)| p)
+            .cloned()
+            .collect();
+
+        // Simple confchange
+        if changes.len() == 1 {
+            match changes.pop().unwrap() {
+                (ConfChangeType::RemoveNode, p) => remove_peers.push(p),
+                (_, p) => add_peers.push(p),
+            }
+        } else {
+            // Joint confchange
+            for (change, mut peer) in changes {
+                match (
+                    find_peer(&region, peer.get_store_id()).map(|p| p.get_role()),
```

Comment:
I suggest to assume role is set properly by caller to make it simple.

Reply:
We can't do that, as before proposing, `check_conf_change` expect `AddNode` confchange should come with peer that has `PeerRole::Voter` instead of `PeerRole::ImcommingVoter`. Also setting the role by the caller will make the test complicated.